shared-bindings/aes/aes.c

```diff
@@ -0,0 +1,355 @@
+  return mp_const_none;
+}
+
+MP_DEFINE_CONST_FUN_OBJ_VAR(aes_rekey_obj, 2, aes_rekey);
+
+
+STATIC mp_obj_t aes_set_mode(mp_obj_t aes_obj, mp_obj_t mode_obj) {
+  if (!MP_OBJ_IS_TYPE(aes_obj, &aes_aes_type)) {
+    mp_raise_TypeError_varg(translate("Expected a %q"), aes_aes_type.name);
+  }
+  // Convert parameters into expected types.
+  aes_obj_t *self = MP_OBJ_TO_PTR(aes_obj);
+
+  int mode = mp_obj_get_int(mode_obj);
+  switch (mode) {
+  case AES_MODE_CBC: /* CBC */
+    break;
+  case AES_MODE_ECB: /* ECB */
+    break;
+  case AES_MODE_CTR: /* CTR */
+    break;
+  default:
+    mp_raise_TypeError(translate("Requested AES mode is unsupported"));
+  }
+
+  common_hal_aes_set_mode(self, mode);
+  return mp_const_none;
+}
+
+MP_DEFINE_CONST_FUN_OBJ_2(aes_set_mode_obj, aes_set_mode);
+
+
+STATIC byte *duplicate_data(const byte *src_buf, size_t len) {
+  byte *dest_buf = m_new(byte, len);
+  memcpy(dest_buf, src_buf, len);
+  return dest_buf;
+}
+
+//|   .. method:: encrypt_in_place(buf)
+//|
+//|      Encrypt the provided buffer in-place without copying it first.
+//|      The buffer must be a multiple of 16 bytes.
+//|
+STATIC mp_obj_t aes_encrypt_in_place(mp_obj_t aes_obj, mp_obj_t buf) {
+  if (!MP_OBJ_IS_TYPE(aes_obj, &aes_aes_type)) {
+    mp_raise_TypeError_varg(translate("Expected a %q"), aes_aes_type.name);
+  }
+  // Convert parameters into expected types.
+  aes_obj_t *aes = MP_OBJ_TO_PTR(aes_obj);
+  mp_buffer_info_t bufinfo;
+  mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_READ);
+  if ((bufinfo.len & (AES_BLOCKLEN - 1)) != 0) {
+    mp_raise_ValueError(translate("Buffer must be a multiple of 16 bytes"));
+  }
+
+  common_hal_aes_encrypt(aes, (uint8_t *)bufinfo.buf, bufinfo.len);
+  return mp_const_none;
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(aes_encrypt_in_place_obj,
+                                 aes_encrypt_in_place);
+
+//|   .. method:: encrypt_into(src, dest)
+//|
+//|      Encrypt the buffer from ``src`` into ``dest``.
+//|      The buffers must be a multiple of 16 bytes, and must
+//|      be equal length
+//|
+STATIC mp_obj_t aes_encrypt_into(mp_obj_t aes_obj, mp_obj_t src,
+                                 mp_obj_t dest) {
+  if (!MP_OBJ_IS_TYPE(aes_obj, &aes_aes_type)) {
+    mp_raise_TypeError_varg(translate("Expected a %q"), aes_aes_type.name);
+  }
+  // Convert parameters into expected types.
+  aes_obj_t *aes = MP_OBJ_TO_PTR(aes_obj);
+
+  mp_buffer_info_t srcbufinfo;
+  mp_get_buffer_raise(src, &srcbufinfo, MP_BUFFER_READ);
+  if ((srcbufinfo.len & (AES_BLOCKLEN - 1)) != 0) {
+    mp_raise_ValueError(translate("Buffer must be a multiple of 16 bytes"));
+  }
+
+  mp_buffer_info_t destbufinfo;
+  mp_get_buffer_raise(dest, &destbufinfo, MP_BUFFER_READ);
+  if (destbufinfo.len != srcbufinfo.len) {
+    mp_raise_ValueError(
+        translate("Source and destination buffers must be the same length"));
+  }
+
+  memcpy(destbufinfo.buf, srcbufinfo.buf, srcbufinfo.len);
```

Comment:
Why do this? Isn't it simpler to have a dual buffer implementation under the hood and pass it twice for in-place?

Reply:
The actual library call takes the form of `AES_CBC_decrypt_buffer(struct AES_ctx* ctx, uint8_t* buf, uint32_t length);` or even `AES_ECB_decrypt(const struct AES_ctx* ctx, uint8_t* buf);`.  Internally it's a single buffer, so we need to do the copy.